### ISSUE TITLE
Can this library be compiled and used under the window operating system？

### ISSUE THREAD

**Issue by JarlenLai** [etcd-cpp-apiv3/etcd-cpp-apiv3#22]
I hope it would be better to explain that this library can be used under those operating systems.

**Comment by sighingnow** [comment]
It should work but I haven't test on windows yet. I will post more details here after do a fully test.

**Comment by emusgrave** [comment]
I am using it successfully on Windows. I had to do a few things to get it working:
1) In Client.cpp, use winsock instead of `sys/socket.h`
```c++
#ifdef _WIN32
#include <winsock2.h>
#include <ws2tcpip.h>
#else
#include <netdb.h>
#include <sys/socket.h>
#endif
```

2) In kv.proto, there are compile errors because the `DELETE` keyword has a conflict on Windows. Instead of trying to figure out how to _really_ solve it, I just renamed the DELETE event on line 34 to EVENT_DELETE. And then I had to make the same change in AsyncWatchResponse.cpp (line 26).
I'm all ears if someone has a more elegant way to solve this on Windows.

3) I personally had many issues with the CMake setup on Windows. I use vcpkg for all of my dependencies, and things didn't go smoothly integrating this library. There were a lot of changes that I made that probably break other platforms, so I have not made a PR for it all. One tip I can give is that I had to copy all of the proto's that come with protobuf into the etcd-cpp-apiv3/proto/google/protobuf folder, so that it could find them properly. 

It might be a good idea to edit the proto files that this repo uses and remove all of the references to what looks like `go`-specific stuff. In other words, I don't really think that the `import "gogoproto/gogo.proto";` is necessary for this library to work. Nor are all of the ```option (google.api.http) = {
        post: "/v3/kv/range"
        body: "*"
    };``` sprinkled throughout the proto files.

**Comment by sighingnow** [comment]
Thanks for noticing that @emusgrave.

- For 1, it is trivial to add correct header files. On windows the `getaddrinfo` call also requires a `WSAStartup` call.
- For 2, I have renamed `DELETE` to `DELETE_` in proto. Seems that there's no easy way to prevent the macro `DELETE` on windows.
- For 3, vcpkg is fine to setup the dependencies, and it might be the only easy way to setup those dependencies, since the `cpprestsdk` library suffers tons of errors with mingw-w64. For the location error of `google/protobuf` files, it was caused by incorrect installation of `protobuf` by vcpkg, where the `protoc` should be placed under `$ROOT/bin/proto`, but it was placed to `$ROOT/tools/grpc/proto`. If I understand it correctly, the relative location does matter for `protoc`.
- For 4, that is because we don't want to maintain our customized proto files, we assume we can copy them directly from the upstream etcd repo, that's why we don't do clean up.
